rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+    bytes ek_key = 3;
+  }
+  // The attestation key and the parameters necessary to remotely verify it as
+  // related to the endorsement key.
+  TPMAttestationParameters attestation_params = 4;
+}
+
+message RegisterUsingTPMMethodRequest {
+  oneof payload {
+    // Initial information sent from the client to the server.
+    RegisterUsingTPMMethodInitialRequest init = 1;
+    // The challenge response required to complete the TPM join process. This is
+    // sent in response to the servers challenge.
+    RegisterUsingTPMMethodEnrollChallengeResponse challenge_response = 2;
+  }
+}
+
+// These values are used by the TPM2.0 `ActivateCredential` command to produce
+// the solution which proves possession of the EK and AK.
+//
+// For a more in-depth description see:
+// - https://pkg.go.dev/github.com/google/go-attestation/attest#EncryptedCredential
+// - https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part3_Commands_code_pub.pdf (Heading 12.5.1 "TPM2_ActivateCredential" "General Description")
+// - https://github.com/google/go-attestation/blob/v0.4.3/attest/activation.go#L199
+// - https://github.com/google/go-tpm/blob/v0.3.3/tpm2/credactivation/credential_activation.go#L61
+message RegisterUsingTPMMethodEnrollChallengeRequest {
+  // The `credential_blob` parameter to be used with the `ActivateCredential`
+  // command. This is used with the decrypted value of `secret` in a
+  // cryptographic process to decrypt the solution.
+  bytes credential_blob = 1;
+  // The `secret` parameter to be used with `ActivateCredential`. This is a
+  // seed which can be decrypted with the EK. The decrypted seed is then used
+  // when decrypting `credential_blob`.
+  bytes secret = 2;
+}
+
+message RegisterUsingTPMMethodResponse {
+  oneof payload {
+    // The challenge required to complete the TPM join process. This is sent to
+    // the client in response to the initial request.
+    RegisterUsingTPMMethodEnrollChallengeRequest challenge_request = 1;
+    // The signed certificates resulting from the join process.
+    Certs certs = 2;
+  }
+}
+```
+
+## Product Considerations
+
+### Licensing
+
+The TPM join method will be restricted to Enterprise/Cloud licensed clusters.
+
+### Improvement: Platform Attestation
+
+Eventually, we can use the TPM to perform a platform attestation to inspect
+the PCR values and event log during the join process. These values can
+be logged and rules can be put in place to block joins where a host's state
+sufficiently deviates from the known-good and expected state.
+
+This work is deferred for now as platform attestation is a complex topic and
+greatly increases the complexity of the implementation and also the complexity
+of using the feature. This can be revisited once there is a clearly defined
+use-case and demonstrated demand.
+
+### Improvement: Storing Bot private key material in the TPM
+
+In addition to using the TPM for the join process, we can also use this to
```

Comment:
From my perspective, we should always fall back to this method.
Even if we use the TPM joining method to guarantee the node is the specific node we want to grant access to, having the credentials stored somewhere - usually on disk - will create a major security consideration where extraaffiliate credentials can be used to gain access to API without being the node itself.
We should store the PK material in the TPM for agents/bots and avoid at all costs writing those credentials anywhere.
By default, all agents will write their credentials to disk and when using TPM they should not. 
This guarantees that any agent will always bootstrap from that node.

Reply:
I heavily agree that we should *eventually* support storing the credentials somewhere more secure than disk - I felt like it was appropriate to leave it out of this RFD to prevent the scope ballooning. There's some interesting technical questions around whether we'd want to integrate directly with the TPM API or use a PKCS11 conversion layer to give us the option of also supporting HSMs. In addition, the way that `tbot` is currently designed means it would be a fairly large refactor to use a HSM or TPM for key storage.

To some extent, when using `tpm` joining, it wouldn't be that awkward to just use the memory backend for storing the Bot's credentials - this at least makes it much more complex to exfiltrate.

Reply:
I agree with the flow for `tbot joinig` storing stuff in memory with some protection to prevent the OS from dumping its contents to swap, but my comment accounts more for the agents' case. The description of the RFD clearly states it will also be applied for stuff besides tbot. 

It's my ultimate goal to be able to use the TPM as the agent's credentials holder, I would do everything to have it early 😆

Reply:
🤣  Yeah - I think the main issue is that key storage in Bots/Agents is completely different - and to some extent, its valuable to have the join method decoupled from the key storage method.

I'm more than happy to put together a different RFD for using TPMs as a key store for both `tbot` and agents, that way there's room for us to compare using the TPM API directly versus treating the TPM as a PKCS11 compatible keystore (which I think we already support).

Reply:
@rosstimothy There's definitely performance concerns here - I started writing a benchmark last night to compare RSA performance in-memory and via the TPM. I'd like to do one that covers TLS handshakes and SSH handshakes as well.

Reply:
TPMs are visibly slow, in human terms. I've seen key creation take in the order of seconds and Windows Hello can be a bit slow as well (which would be a sign operation). Not end-of-the-world slow, but good to keep in mind.